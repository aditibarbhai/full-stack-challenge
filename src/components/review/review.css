.review-form-container {
    width: 75%;
    margin: auto;
    padding: 3rem;
    height: 100vh;
}

.review-form-container .employee-info {
    text-align: center;
}

.review-form-container .employee-info > * {
    margin-bottom: 1rem;
}

.image-frame {
    padding: 1rem;
    border: 1px solid #dadada;
    display: inline-block;
}

.review-action-buttons {
    display: flex;
    justify-content: space-evenly;
    color: white;
    margin-top: 20px;
}

.review-action-buttons > * {
    min-width: 48px;
}

.button-send {
    background: linear-gradient(45deg, #FE6B8B 30%, #FF8E53 90%);
    border-radius: 3px;
    border: 0;
    height: 48px;
    padding: 0 30px;
    box-shadow: 0 3px 5px 2px rgba(255, 105, 135, .3);
}

.button-send .icon-send {
    margin-left: 8px;
}

.button-save {
    background: linear-gradient(45deg, #2196F3 30%, #21CBF3 90%);
    border-radius: 3px;
    border: 0;
    height: 48px;
    padding: 0 30px;
    box-shadow: 0 3px 5px 2px rgba(33, 203, 243, .3);
}

.button-save .icon-save {
    margin-right: 8px;
}

/*.button-clear-all {*/
    /*background: #b9c5c7;*/
    /*border-radius: 3px;*/
    /*border: 0;*/
    /*height: 48px;*/
    /*padding: 0 30px;*/
    /*box-shadow: 0 3px 5px 2px rgba(33, 203, 243, .3);*/
/*}*/

.button-clear-all .icon-clear-all {
    margin-right: 8px;
}
