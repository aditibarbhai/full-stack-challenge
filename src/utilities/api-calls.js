const apiEndpoint = 'http://localhost:3001';

export const getEmployees = () =>
    fetch(`${apiEndpoint}/employees`, { method: 'GET' })
    .then(response => response.json());

export const getEmployee = (id) =>
    fetch(`${apiEndpoint}/employees/${id}`, { method: 'GET' })
        .then(response => response.json());

export const getToDos = (id) =>
    fetch(`${apiEndpoint}/todo?employee=${id}&_sort=completed`, { method: 'GET' })
        .then(response => response.json());

export const updateTodo = (todo) =>
    fetch(`${apiEndpoint}/todo/${todo.id}`, {
        method: 'PUT',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(todo)
    }).then(response => response.json());

export const getReviews = (employeeId) =>
    fetch(`${apiEndpoint}/reviews/${employeeId}`, { method: 'GET' })
        .then(response => response.json());

export const addReview = (review) =>
    fetch(`${apiEndpoint}/reviews`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(review)
    }).then(response => response.json());

export const updateReview = (review) =>
    fetch(`${apiEndpoint}/reviews/${review.id}`, {
        method: 'PUT',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(review)
    }).then(response => response.json());