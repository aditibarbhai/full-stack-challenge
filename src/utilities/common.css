body {
    margin: 0;
    padding: 0;
    font-family: sans-serif;
}

#root {
    background-color: #2bbcf4;
    width: 100vw;
    min-height: 100vh;
}

.truncate {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}