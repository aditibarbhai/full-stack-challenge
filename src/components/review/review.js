import React, { Component } from 'react';
import Paper from '@material-ui/core/Paper';
import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import SendIcon from '@material-ui/icons/Send'
import SaveIcon from '@material-ui/icons/Save';
import ArrowBackIcon from '@material-ui/icons/ArrowBack';

import { withStyles } from '@material-ui/core/styles';
import { addReview, updateReview, updateTodo } from '../../utilities/api-calls';

import './review.css';

const styles = {
    backButton: {
        background: 'white',
        boxShadow: 'none',
        color: 'black',
        '&:hover': {
            color: 'white',
            background: '#dadada'
        }
    }
};

export class Review extends Component {
    constructor(props) {
        super(props);

        this.state = {
            id: '',
            connection: '',
            review: '',
            reviewer: props.location.state ? props.location.state.reviewer : {},
            reviewee: props.location.state ? props.location.state.todo.toReview : {},
            todo: props.location.state ? props.location.state.todo : {},
        };

        this.updateConnectionText = this.updateConnectionText.bind(this);
        this.updateReviewText = this.updateReviewText.bind(this);
        this.handleSaveClick = this.handleSaveClick.bind(this);
        this.handleSendClick = this.handleSendClick.bind(this);
    }

    componentDidMount() {
        const { match } = this.props;
        // getEmployee(match.params.id).then(reviewee => this.setState({ reviewee }));
    }

    updateConnectionText(event) {
        console.log(event.target.value);
        this.setState({ connection: event.target.value });
    }

    updateReviewText(event) {
        this.setState({ review: event.target.value });
    }

    submitReview(published) {
        let review = {
            connection: this.state.connection,
            review: this.state.review,
            reviewer: this.state.reviewer.id,
            reviewee: this.state.reviewee.id,
            published
        };

        if (this.state.id) {
            review.id = this.state.id;
        }

        if (!this.state.id) {
            // create new entry for this review
            return addReview(review);
        } else {
            // update review in DB
            return updateReview(review);
        }
    }

    handleSaveClick() {
        this.submitReview(false).then(response => {
            this.setState({ id: response.id });
        });
    }

    handleSendClick(event) {
        event.preventDefault();

        this.submitReview(true).then(response => {
            this.setState({ id: response.id });

            let updatedToDo = {
                ...this.state.todo,
                completed: true
            };

            // update "todo" and set it as done
            updateTodo(updatedToDo).then(response => {
                // navigate back to home page
                this.props.history.push('/');
            });
        });
    }

    render() {
        return (
            <Paper className="review-form-container">
                <Grid container spacing={16}>
                    <Grid className="navigation" item xs={12}>
                        <Button
                            onClick={() => this.props.history.push('/')}
                            variant="fab" color="primary"
                            aria-label="back"
                            className={this.props.classes.backButton}
                        >
                            <ArrowBackIcon className="icon-back" />
                        </Button>
                    </Grid>
                    <Grid className="employee-info" item md={3} xs={12}>
                        <div className="image-frame">
                            <img src={this.state.reviewee.avatar} alt={`${this.state.reviewee.firstName} ${this.state.reviewee.lastName}`} />
                        </div>
                        <Typography variant="title"> { `${this.state.reviewee.firstName} ${this.state.reviewee.lastName}` } </Typography>
                    </Grid>
                    <Grid item md={9} xs={12} className="form">
                        <form>
                            <TextField
                                id="connection"
                                label={`How do you know ${this.state.reviewee.firstName}?`}
                                value={this.state.connection}
                                margin="normal"
                                onChange={this.updateConnectionText}
                                fullWidth
                            />
                            <TextField
                                id="feedback"
                                label={`What feedback would you like to provide ${this.state.reviewee.firstName}?`}
                                value={this.state.review}
                                margin="normal"
                                rows={3}
                                rowsMax={10}
                                onChange={this.updateReviewText}
                                fullWidth
                                multiline
                            />

                            <div className="review-action-buttons">
                                <Button
                                    variant="contained"
                                    color="primary"
                                    className="button-save"
                                    onClick={this.handleSaveClick}
                                    disabled={(!this.state.connection && !this.state.review)}
                                >
                                    <SaveIcon className="icon-save" />
                                    Save
                                </Button>

                                <Button
                                    variant="contained"
                                    type="submit"
                                    color="primary"
                                    className="button-send"
                                    onClick={this.handleSendClick}
                                    disabled={(!this.state.connection || !this.state.review)}
                                >
                                    Send
                                    <SendIcon className="icon-send" />
                                </Button>
                            </div>
                        </form>
                    </Grid>
                </Grid>
            </Paper>
        );
    }
}

export default withStyles(styles)(Review);
