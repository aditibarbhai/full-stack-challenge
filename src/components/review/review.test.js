import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { Review } from './review';

jest.mock('../../utilities/api-calls.js');
import * as apiCalls from '../../utilities/api-calls';

describe('Review tests', () => {
    let props = {};

    beforeEach(() => {
        props = {
            location: {
                state: {
                    reviewer: {
                        id: 1
                    },
                    todo: {
                        toReview: {
                            id: 2
                        }
                    }
                }
            },
            classes: {
                backButton: ''
            }
        };
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it ('should render without errors', () => {
        const tree = renderer
            .create(<Review {...props} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it ('updateConnectionText should update state.connection to value of user input', () => {
        const wrapper = shallow(<Review {...props} />);

        wrapper.instance().updateConnectionText({ target: { value: 'test' } });

        expect(wrapper.state().connection).toBe('test');
    });

    it ('updateReviewText should update state.review to value of user input', () => {
        const wrapper = shallow(<Review {...props} />);

        wrapper.instance().updateReviewText({ target: { value: 'test review' } });

        expect(wrapper.state().review).toBe('test review');
    });

    it ('submitReview call addReview if state.id is falsy', () => {
        const wrapper = shallow(<Review {...props} />),
            newReview = {
                connection: '',
                review: '',
                reviewer: props.location.state.reviewer.id,
                reviewee: props.location.state.todo.toReview.id,
                published: false
            };

        wrapper.instance().submitReview(false);

        expect(apiCalls.addReview).toHaveBeenCalledWith(newReview);
    });

    it ('submitReview call updateReview if state.id is defined', () => {
        const wrapper = shallow(<Review {...props} />),
            newReview = {
                id: 3,
                connection: '',
                review: '',
                reviewer: props.location.state.reviewer.id,
                reviewee: props.location.state.todo.toReview.id,
                published: true
            };

        wrapper.setState({ id: 3 });

        wrapper.instance().submitReview(true);

        expect(apiCalls.updateReview).toHaveBeenCalledWith(newReview);
    });
});