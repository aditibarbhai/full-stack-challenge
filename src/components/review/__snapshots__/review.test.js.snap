// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review tests should render without errors 1`] = `
<div
  className="MuiPaper-root-1 MuiPaper-elevation2-5 MuiPaper-rounded-2 review-form-container"
>
  <div
    className="MuiGrid-container-28 MuiGrid-spacing-xs-16-50"
  >
    <div
      className="MuiGrid-item-29 MuiGrid-grid-xs-12-67 navigation"
    >
      <button
        aria-label="back"
        className="MuiButtonBase-root-148 MuiButton-root-124 MuiButton-fab-139 MuiButton-contained-133 MuiButton-containedPrimary-134 MuiButton-raised-136 MuiButton-raisedPrimary-137"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-125"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-151 icon-back"
            color={undefined}
            focusable="false"
            viewBox="0 0 24 24"
          >
            <g>
              <path
                d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              />
            </g>
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-158"
        />
      </button>
    </div>
    <div
      className="MuiGrid-item-29 MuiGrid-grid-xs-12-67 MuiGrid-grid-md-3-86 employee-info"
    >
      <div
        className="image-frame"
      >
        <img
          alt="undefined undefined"
          src={undefined}
        />
      </div>
      <h2
        className="MuiTypography-root-165 MuiTypography-title-171"
      >
         
        undefined undefined
         
      </h2>
    </div>
    <div
      className="MuiGrid-item-29 MuiGrid-grid-xs-12-67 MuiGrid-grid-md-9-92 form"
    >
      <form>
        <div
          aria-describedby={undefined}
          className="MuiFormControl-root-189 MuiFormControl-marginNormal-190 MuiFormControl-fullWidth-192"
        >
          <label
            className="MuiFormLabel-root-198 MuiInputLabel-root-193 MuiInputLabel-formControl-194 MuiInputLabel-animated-197"
            data-shrink={false}
            htmlFor="connection"
          >
            How do you know undefined?
          </label>
          <div
            className="MuiInput-root-203 MuiInput-fullWidth-210 MuiInput-formControl-204 MuiInput-underline-207"
          >
            <input
              aria-invalid={false}
              autoComplete={undefined}
              autoFocus={undefined}
              className="MuiInput-input-211"
              defaultValue={undefined}
              disabled={false}
              id="connection"
              name={undefined}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={undefined}
              onKeyUp={undefined}
              placeholder={undefined}
              readOnly={undefined}
              required={undefined}
              rows={undefined}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          aria-describedby={undefined}
          className="MuiFormControl-root-189 MuiFormControl-marginNormal-190 MuiFormControl-fullWidth-192"
        >
          <label
            className="MuiFormLabel-root-198 MuiInputLabel-root-193 MuiInputLabel-formControl-194 MuiInputLabel-animated-197"
            data-shrink={false}
            htmlFor="feedback"
          >
            What feedback would you like to provide undefined?
          </label>
          <div
            className="MuiInput-root-203 MuiInput-fullWidth-210 MuiInput-formControl-204 MuiInput-multiline-209 MuiInput-underline-207"
          >
            <div
              className="Textarea-root-216"
              style={
                Object {
                  "height": 57,
                }
              }
            >
              <textarea
                aria-hidden="true"
                className="Textarea-shadow-218 Textarea-textarea-217"
                readOnly={true}
                rows="1"
                tabIndex={-1}
                value=""
              />
              <textarea
                aria-hidden="true"
                className="Textarea-shadow-218 Textarea-textarea-217"
                defaultValue={undefined}
                readOnly={true}
                rows={3}
                tabIndex={-1}
                value=""
              />
              <textarea
                aria-invalid={false}
                autoComplete={undefined}
                autoFocus={undefined}
                className="Textarea-textarea-217 MuiInput-input-211 MuiInput-inputMultiline-213"
                defaultValue={undefined}
                disabled={false}
                id="feedback"
                name={undefined}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={undefined}
                onKeyUp={undefined}
                placeholder={undefined}
                readOnly={undefined}
                required={undefined}
                rows={3}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          className="review-action-buttons"
        >
          <button
            className="MuiButtonBase-root-148 MuiButtonBase-disabled-149 MuiButton-root-124 MuiButton-contained-133 MuiButton-containedPrimary-134 MuiButton-raised-136 MuiButton-raisedPrimary-137 MuiButton-disabled-142 button-save"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="-1"
            type="button"
          >
            <span
              className="MuiButton-label-125"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-151 icon-save"
                color={undefined}
                focusable="false"
                viewBox="0 0 24 24"
              >
                <g>
                  <path
                    d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
                  />
                </g>
              </svg>
              Save
            </span>
          </button>
          <button
            className="MuiButtonBase-root-148 MuiButtonBase-disabled-149 MuiButton-root-124 MuiButton-contained-133 MuiButton-containedPrimary-134 MuiButton-raised-136 MuiButton-raisedPrimary-137 MuiButton-disabled-142 button-send"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="-1"
            type="submit"
          >
            <span
              className="MuiButton-label-125"
            >
              Send
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-151 icon-send"
                color={undefined}
                focusable="false"
                viewBox="0 0 24 24"
              >
                <g>
                  <path
                    d="M2.01 21L23 12 2.01 3 2 10l15 2-15 2z"
                  />
                </g>
              </svg>
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
